### PR TITLE
#182201659 [ configure-postgres-and-sequelize-ORM ]

### PR DESCRIPTION
**Pull request by Placide123**
#### What does this PR do?

- Configure BareFoot Nomad to use PostgreSQL as DB Engine and Sequelize ORM

#### Description of Task to be completed?

- User of an application can be able to send data into the database and use those data.

#### How should this be manually tested?

- Clone the repo

- checkout to a branch called  `chore-configure-postgres-and-sequelize-ORM-#182201659`

- Run `npm install` to install modules

- Add `.env` file to set environment variable as given `.env.example`.

- Run `npm start:dev` to start the development server

- use postman as a tool for testing, get data by `localhost:3000/api/v1/users/` and create user by post `localhost:3000/api/v1/users/create`

#### What are the relevant Pivotal Tracker/Trello stories?

- https://www.pivotaltracker.com/n/projects/2569105

#### Questions:
- N/A.

### CODE REVIEW

**Review by semy-tuyi — DISMISSED** [review]

**Review by uwambajeddy — COMMENTED** [review]
src/models/index.js

All the files are supposed be in Es6

**Review by semy-tuyi — DISMISSED** [review]

**Review by semy-tuyi — DISMISSED** [review]

**Review by uwambajeddy — DISMISSED** [review]
LGTM, just a quick note you didn't suggest that they should remember to add ENVIRONMENT VARIABLES and follow instructions written in a ".env.example" file

File: src/config/config.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+
+import dotenv from 'dotenv';
+
+dotenv.config(".env");
```
Comment (didasmbalanya) [inline src/config/config.js]:
```suggestion
import "dotenv/config"
```

Both lines can be done with ones

File: src/config/config.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+//     },
+//     test: {
+//       username: PGUSER,
+//       password: PGPASSWORD,
+//       database: TEST_PGDATABASE,
+//       host: PGHOST,
+//       port: PGPORT,
+//       dialect: 'postgres',
+//       logging: false
+//     },
```
Comment (didasmbalanya) [inline src/config/config.js]: remove commented code

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+					message:"Retrieved"});
+			})
+		
+	}
+	catch(error) {
+		return res.status(500).json(error.message);
+	}
+	
+};
+
```
Comment (didasmbalanya) [inline src/controllers/user.controller.js]: this is code that should be on the `user.service.js` file. Code that interact with an external service is in the services not controller

File: src/index.js [hunk 4]
```diff
@@ -1,6 +1,27 @@
...
+
+
+app.use(cors());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(express.json()); 
+
+
+/* ========== Start:: User api url ========= */ 
+  app.use('/api/v1/users', usersRoutes);
+/* ============== Start:: User api ========= */
```
Comment (didasmbalanya) [inline src/index.js]: remove this and put it inside a router/index.ts file.
That fill will export one router which will be called here as. 
```suggestion
  app.use(router);
```

File: src/index.js [hunk 5]
```diff
@@ -1,6 +1,27 @@
...
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(express.json()); 
+
+
+/* ========== Start:: User api url ========= */ 
+  app.use('/api/v1/users', usersRoutes);
+/* ============== Start:: User api ========= */ 
+
+app.listen(PORT, () => {
+  app.emit("Started")
```
Comment (didasmbalanya) [inline src/index.js]: remove

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

**Review by alaintresor — DISMISSED** [review]

**Review by semy-tuyi — DISMISSED** [review]

**Review by Patricknshimiyimana — DISMISSED** [review]
This looks good the only optional thing remaining is squashing these commits into one.

File: src/routes/user.routes.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import express from 'express';
+import { createUser, getAllUsers } from "../services/user.service.js";
+
+const router = express.Router();
+
+    router.get('/', getAllUsers); 
+    router.post('/create', createUser);
```
Comment (uwambajeddy) [inline src/routes/user.routes.js:7]: I hope you know that I will overwrite this end points.

update folder name `routes` to `routers`

also add this file `src/routers/user.routes.js`  to `src/routers/api/user.routes.js`

File: src/app.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+const app = express();
+
+app.use(cors());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(express.json()); 
+app.use(router);
+
+app.use('/',(req, res) => {
+    res.status(200).json({ success: true, message: "You are using Barefoot nomad app." })
+  });
```
Comment (uwambajeddy) [inline src/app.js]: No end point are allowed to be in defined in `src\app.js` file. They will be defined in the `routers\index.js` file

File: src/database/models/users.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+     */
+    static associate(models) {
+      // define association here
+    }
+  }
+  users.init({
+    fullname: DataTypes.STRING,
+    username: DataTypes.STRING,
+    email: DataTypes.STRING,
+    role: DataTypes.STRING,
```
Comment (uwambajeddy) [inline src/database/models/users.js:18]: there must be specific values for `role field`

File: src/database/migrations/20220523191126-create-users.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      allowNull: false
+    },
+    email: {
+      type: Sequelize.STRING,
+      allowNull: false
+    },
+    role: {
+      type: Sequelize.STRING,
+      allowNull: false
+    },
```
Comment (uwambajeddy) [inline src/database/migrations/20220523191126-create-users.js:25]: also here you must specify specific values

File: src/database/models/users.js (hunk 8 above)
Comment (Placide123) [reply src/database/models/users.js:18]: hi, thank you for the feedback but in the task they precise to say I have to put dummy data in the database that's why I put that one.

File: src/app.js (hunk 7 above)
Comment (Placide123) [reply src/app.js]: let me do it

**Review by HiHeritier — DISMISSED** [review]

**Review by semy-tuyi — DISMISSED** [review]

**Review by joshwambere — CHANGES_REQUESTED** [review]
@Placide123   as you saw the history on develop we have two folder `Routers` and Routes which technically should be one folder so find a ways to merge the into one folder

File: src/services/user.service.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+			})
+		
+	}
+	catch(error) {
+		return res.status(500).json(error.message);
+	}
+	
+};
+
+export const createUser = async (req, res) => {
```
Comment (joshwambere) [inline src/services/user.service.js:19]: try linting cause spacing here is a mess

File: src/services/user.service.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import db from '../database/models/index.js'
+const users = db['users']
+
+export const getAllUsers = async (req, res) => {
+ 
+	try{
+			users.findAndCountAll().then(users => {
+				
+				return res.status(200).json({
+					status:"success",
```
Comment (joshwambere) [inline src/services/user.service.js]: status may be you can use success set to a boolean `true or false`

**Review by joshwambere — CHANGES_REQUESTED** [review]

**Review by semy-tuyi — APPROVED** [review]

**Review by Patricknshimiyimana — APPROVED** [review]

**Review by joshwambere — APPROVED** [review]

**Pull request merged**